File: exp-04/q1-max-min-divide-conquer/solution.js
const readline = require('readline');

const rl = readline.createInterface({ input: process.stdin, output: process.stdout });

function findMinMax(a, lo, hi) {
    if (lo === hi) return [a[lo], a[lo]];
    if (hi === lo + 1) return a[lo] < a[hi] ? [a[lo], a[hi]] : [a[hi], a[lo]];
    const mid = Math.floor((lo + hi) / 2);
    const [minL, maxL] = findMinMax(a, lo, mid);
    const [minR, maxR] = findMinMax(a, mid + 1, hi);
    return [Math.min(minL, minR), Math.max(maxL, maxR)];
}

let lines = [];
rl.on('line', (line) => {
    lines.push(line.trim());
    if (lines.length === 2) {
        const n = parseInt(lines[0], 10);
        const a = lines[1].split(/\s+/).map(Number);
        const [min, max] = findMinMax(a, 0, n - 1);
        console.log(`Min: ${min}`);
        console.log(`Max: ${max}`);
        rl.close();
    }
});

File: exp-04/q1-max-min-divide-conquer/solution.py
def find_min_max(a, lo, hi):
    if lo == hi:
        return a[lo], a[lo]
    if hi == lo + 1:
        return (a[lo], a[hi]) if a[lo] < a[hi] else (a[hi], a[lo])
    mid = (lo + hi) // 2
    min_l, max_l = find_min_max(a, lo, mid)
    min_r, max_r = find_min_max(a, mid + 1, hi)
    return min(min_l, min_r), max(max_l, max_r)


n = int(input())
a = list(map(int, input().split()))
min_val, max_val = find_min_max(a, 0, n - 1)
print(f"Min: {min_val}")
print(f"Max: {max_val}")